File: src/index/store.ts
import Database from 'better-sqlite3';

export interface StoredEmbedding {
    id: string;
    text: string;
    vector: number[];
    model: string;
}

export interface QueryResult {
    id: string;
    text: string;
    score: number;
}

export class EmbeddingStore {
    private db: Database.Database;

    constructor(path: string = 'embeddings.db') {
        this.db = new Database(path);
        this.initialize();
    }

    private initialize(): void {
        this.db.exec(`
            CREATE TABLE IF NOT EXISTS embeddings (
                id TEXT PRIMARY KEY,
                text TEXT NOT NULL,
                vector TEXT NOT NULL,
                model TEXT NOT NULL
            );
        `);
    }

    /**
     * Insert or update an embedding.
     */
    upsert(embedding: StoredEmbedding): void {
        const stmt = this.db.prepare(`
            INSERT INTO embeddings (id, text, vector, model)
            VALUES (@id, @text, @vector, @model)
            ON CONFLICT(id) DO UPDATE SET
                text = excluded.text,
                vector = excluded.vector,
                model = excluded.model;
        `);

        stmt.run({
            id: embedding.id,
            text: embedding.text,
            vector: JSON.stringify(embedding.vector),
            model: embedding.model,
        });
    }

    /**
     * Retrieve top-K most similar embeddings.
     */
    query(
        queryVector: number[],
        options?: {
            limit?: number;
            minScore?: number;
            model?: string;
        }
    ): QueryResult[] {
        const limit = options?.limit ?? 5;
        const minScore = options?.minScore ?? 0;
        const model = options?.model;

        const rows = model
            ? this.db
                  .prepare(
                      'SELECT id, text, vector FROM embeddings WHERE model = ?'
                  )
                  .all(model)
            : this.db
                  .prepare(
                      'SELECT id, text, vector FROM embeddings'
                  )
                  .all();

        const results: QueryResult[] = [];

        for (const row of rows) {
            const vector = JSON.parse(row.vector) as number[];

            const score = cosineSimilarity(queryVector, vector);

            if (score >= minScore) {
                results.push({
                    id: row.id,
                    text: row.text,
                    score,
                });
            }
        }

        return results
            .sort((a, b) => b.score - a.score)
            .slice(0, limit);
    }

    /**
     * Remove an embedding.
     */
    delete(id: string): void {
        this.db
            .prepare('DELETE FROM embeddings WHERE id = ?')
            .run(id);
    }

    /**
     * Clear the entire store.
     */
    clear(): void {
        this.db.exec('DELETE FROM embeddings');
    }
}

/**
 * Cosine similarity between vectors.
 */
function cosineSimilarity(a: number[], b: number[]): number {
    let dot = 0;
    let normA = 0;
    let normB = 0;

    for (let i = 0; i < a.length; i++) {
        dot += a[i] * b[i];
        normA += a[i] * a[i];
        normB += b[i] * b[i];
    }

    if (normA === 0 || normB === 0) {
        return 0;
    }

    return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

File: rust/rlm/src/index/store.rs
#![allow(dead_code)]

#[derive(Debug, Clone)]
pub struct StoredEmbedding {
    pub id: String,
    pub text: String,
    pub vector: Vec<f32>,
    pub model: String,
}

#[derive(Debug, Clone)]
pub struct QueryResult {
    pub id: String,
    pub text: String,
    pub score: f32,
}

/// A simple in-memory embedding store.
#[derive(Default, Debug)]
pub struct EmbeddingStore {
    items: Vec<StoredEmbedding>,
}

impl EmbeddingStore {
    pub fn new() -> Self {
        Self { items: Vec::new() }
    }

    /// Insert or update an embedding.
    pub fn upsert(&mut self, embedding: StoredEmbedding) {
        if let Some(existing) = self.items.iter_mut().find(|e| e.id == embedding.id) {
            *existing = embedding;
        } else {
            self.items.push(embedding);
        }
    }

    /// Retrieve top-K most similar embeddings.
    pub fn query(
        &self,
        query_vector: &[f32],
        limit: Option<usize>,
        min_score: Option<f32>,
        model: Option<&str>,
    ) -> Vec<QueryResult> {
        let limit = limit.unwrap_or(5);
        let min_score = min_score.unwrap_or(0.0);

        let iter = self.items.iter().filter(|item| match model {
            Some(m) => item.model == m,
            None => true,
        });

        let mut results: Vec<QueryResult> = iter
            .filter_map(|item| {
                let score = cosine_similarity(query_vector, &item.vector);
                if score >= min_score {
                    Some(QueryResult {
                        id: item.id.clone(),
                        text: item.text.clone(),
                        score,
                    })
                } else {
                    None
                }
            })
            .collect();

        results.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));
        results.truncate(limit);
        results
    }

    pub fn delete(&mut self, id: &str) {
        self.items.retain(|item| item.id != id);
    }

    pub fn clear(&mut self) {
        self.items.clear();
    }
}

fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    let mut dot = 0.0_f32;
    let mut norm_a = 0.0_f32;
    let mut norm_b = 0.0_f32;

    let len = a.len().min(b.len());
    for i in 0..len {
        dot += a[i] * b[i];
        norm_a += a[i] * a[i];
        norm_b += b[i] * b[i];
    }

    if norm_a == 0.0 || norm_b == 0.0 {
        return 0.0;
    }

    dot / (norm_a.sqrt() * norm_b.sqrt())
}